Material: Fix copy method to deep-clone userData and safely bind customProgramCacheKey*

Subham-KRLX opened the pull request:
Related issue: N/A

**Description**

The `Material.copy()` method did not deep-clone the `userData` property, leading to shared references between cloned materials.  
This caused unexpected behavior when modifying `userData` on one material instance changes would unintentionally affect others.

### Changes made
- Deep clone of `userData` using `structuredClone` (with JSON fallback).
- Safe handling and rebinding of `customProgramCacheKey` to avoid prototype shadowing.
- Preservation of `onBeforeCompile` references during copy operations.

## Benifits
- Prevents shared `userData` references between cloned materials, avoiding cross-instance state bugs.  
- Ensures each cloned material behaves independently with safely bound methods.  
- Improves cloning reliability and consistency with other Three.js material classes.

Commit:
Refactor userData handling in Material.js

mrdoob commented:
I think this is interesting:

```js
			if ( typeof structuredClone === 'function' ) {

				this.userData = structuredClone( source.userData );

			} else {

				this.userData = JSON.parse( JSON.stringify( source.userData ) );

			}
```

The other changes sound like are specific to your project?

Commit:
Adopt structuredClone for userData deep-cloning

- Updated Material.copy() to use structuredClone with JSON fallback
- Updated Object3D.copy() to use structuredClone with JSON fallback
- Updated Texture.copy() to use structuredClone with JSON fallback
- Updated AnimationClip.clone() to use structuredClone with JSON fallback
- Removed customProgramCacheKey handling to keep PR focused

Commit:
Merge remote-tracking branch 'upstream/dev' into material-fix-copy-method-clone-userdata

Commit:
Update screenshots for postprocessing examples after upstream merge

Commit:
Use upstream screenshot for webgl_postprocessing_glitch

Mugen87 commented:
Do you mind rebasing the PR from dev? There are some issues with E2E screenshots that have already been solved via #32224.

Mugen87 commented on src/core/Object3D.js:
```diff
@@ -1567,7 +1567,15 @@ class Object3D extends EventDispatcher {
 		this.animations = source.animations.slice();
 
-		this.userData = JSON.parse( JSON.stringify( source.userData ) );
+		if ( typeof structuredClone === 'function' ) {
```
Is it okay to implement an approach that produces different results based on the support of `structuredClone`?

I agree with the concerns in https://github.com/mrdoob/three.js/pull/25535#issuecomment-1550606611.

Subham-KRLX replied:
Thanks for the context, @Mugen87!

Yes, I’ve gone through #25535 and #26062 those issues mainly occurred because structuredClone wasn’t universally available leading to runtime errors.

In this PR, I’ve handled that by keeping a safe fallback to JSON.parse(JSON.stringify(...)), so it should avoid those earlier problems.

I’ve also rebased from dev as requested to include the latest E2E screenshot fixes from #32224

Commit:
Merge branch 'mrdoob:dev' into material-fix-copy-method-clone-userdata

Mugen87 commented:
Sorry, but I think the PR can't be merged. It is not good if copy methods produce different results on based on the support of  `structuredClone()`. This has been mentioned in https://github.com/mrdoob/three.js/pull/25535#issuecomment-1550606611 as well.

Besides, using `structuredClone()` caused runtime errors on app level that obviously not resolved by this PR.
